packages/url_launcher/url_launcher_platform_interface/lib/link.dart

```diff
@@ -0,0 +1,97 @@
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:ui';
+
+import 'package:flutter/material.dart';
+import 'package:flutter/services.dart';
+
+/// Signature for a function provided by the [Link] widget that instructs it to
+/// follow the link.
+typedef FollowLink = Future<void> Function();
+
+/// Signature for a builder function passed to the [Link] widget to construct
+/// the widget tree under it.
+typedef LinkWidgetBuilder = Widget Function(
+  BuildContext context,
+  FollowLink followLink,
+);
+
+/// Signature for a delegate function to build the [Link] widget.
+typedef LinkDelegate = Widget Function(LinkInfo linkWidget);
+
+final MethodCodec _codec = const JSONMethodCodec();
+
+/// Defines where a Link URL should be open.
+///
+/// This is a class instead of an enum to allow future customizability e.g.
+/// opening a link in a specific iframe.
+class LinkTarget {
+  /// Const private constructor with a [debugLabel] to allow the creation of
+  /// multiple distinct const instances.
+  const LinkTarget._({this.debugLabel});
+
+  /// Used to distinguish multiple const instances of [LinkTarget].
+  final String debugLabel;
+
+  /// Use the default target for each platform.
+  ///
+  /// On Android, the default is [blank]. On the web, the default is [self].
+  ///
+  /// iOS, on the other hand, defaults to [self] for web URLs, and [blank] for
+  /// non-web URLs.
+  static const defaultTarget = LinkTarget._(debugLabel: 'defaultTarget');
+
+  /// On the web, this opens the link in the same tab where the flutter app is
+  /// running.
+  ///
+  /// On Android and iOS, this opens the link in a webview within the app.
+  static const self = LinkTarget._(debugLabel: 'self');
+
+  /// On the web, this opens the link in a new tab or window (depending on the
+  /// browser and user configuration).
+  ///
+  /// On Android and iOS, this opens the link in the browser or the relevant
+  /// app.
+  static const blank = LinkTarget._(debugLabel: 'blank');
+}
+
+/// Encapsulates all the information necessary to build a Link widget.
+abstract class LinkInfo {
+  /// Called at build time to construct the widget tree under the link.
+  LinkWidgetBuilder get builder;
+
+  /// The destination that this link leads to.
+  Uri get uri;
+
+  /// The target indicating where to open the link.
+  LinkTarget get target;
+
+  /// Whether the link is disabled or not.
+  bool get isDisabled;
+}
+
+/// Pushes the [routeName] into Flutter's navigation system via a platform
+/// message.
+Future<ByteData> pushRouteNameToFramework(
+  BuildContext context,
+  String routeName,
+) {
+  final bool isUsingRouter =
+      context.findAncestorWidgetOfExactType<Router>() != null;
+  final MethodCall call = isUsingRouter
+      ? MethodCall('pushRouteInformation', <dynamic, dynamic>{
+          'location': routeName,
+          'state': null,
+        })
+      : MethodCall('pushRoute', routeName);
```

Comment:
This synthesizes the engine sends the message to framework to push the new route. For navigator, this should be sufficient, but for router, it will think this push route information is coming from the engine, so it will not send routeInformationUpdate to the engine. you will still need to update the url/browser history entry by sending the routeInformationUpdated

Reply:
I can confirm the behavior you are describing for Navigator. I haven't tried the Router yet. Is this a good app to test with (https://github.com/flutter/flutter/pull/63424/) ?

Reply:
yes, that app should be a good example